Firestore/core/test/firebase/firestore/local/counting_query_engine.h

```diff
@@ -0,0 +1,204 @@
+#include "Firestore/core/src/firebase/firestore/local/remote_document_cache.h"
+#include "Firestore/core/src/firebase/firestore/model/document_key_set.h"
+#include "Firestore/core/src/firebase/firestore/model/document_map.h"
+#include "Firestore/core/src/firebase/firestore/model/snapshot_version.h"
+
+namespace firebase {
+namespace firestore {
+namespace local {
+
+class LocalDocumentsView;
+class WrappedMutationQueue;
+class WrappedRemoteDocumentCache;
+
+/**
+ * A test-only QueryEngine that forwards all API calls and exposes the number of
+ * documents and mutations read.
+ */
+class CountingQueryEngine : public QueryEngine {
+ public:
+  CountingQueryEngine(QueryEngine* query_engine, bool index_free)
+      : query_engine_(query_engine), index_free_(index_free) {
+  }
+
+  void ResetCounts();
+
+  void SetLocalDocumentsView(LocalDocumentsView* local_document) override;
+
+  model::DocumentMap GetDocumentsMatchingQuery(
+      const core::Query& query,
+      const model::SnapshotVersion& last_limbo_free_snapshot_version,
+      const model::DocumentKeySet& remote_keys) override;
+
+  /**
+   * Returns whether the backing query engine is optimized to perform key-based
+   * lookups.
+   */
+  // TODO(mrschmidt): Come up with a name that describes the behavior change
+  bool is_index_free() {
+    return index_free_;
+  }
+
+  /**
+   * Returns the number of documents returned by the RemoteDocumentCache's
+   * `GetMatching()` API (since the last call to `ResetCounts()`)
+   */
+  int documents_read_by_query() {
+    return documents_read_by_query_;
+  }
+
+  /**
+   * Returns the number of documents returned by the RemoteDocumentCache's
+   * `Get()` and `GetAll()` APIs (since the last call to `ResetCounts()`)
+   */
+  int documents_read_by_key() {
+    return documents_read_by_key_;
+  }
+
+  /**
+   * Returns the number of mutations returned by the MutationQueue's
+   * `getAllMutationBatchesAffectingQuery()` API (since the last call to
+   * `ResetCounts()`)
+   */
+  int mutations_read_by_query() {
+    return mutations_read_by_query_;
+  }
+
+  /**
+   * Returns the number of mutations returned by the MutationQueue's
+   * `AllMutationBatchesAffectingDocumentKey()` and
+   * `AllMutationBatchesAffectingDocumentKeys()` APIs (since the last call to
+   * `ResetCounts()`)
+   */
+  int mutations_read_by_key() {
+    return mutations_read_by_key_;
+  }
+
+ private:
+  friend class WrappedMutationQueue;
+  friend class WrappedRemoteDocumentCache;
+
+  QueryEngine* query_engine_;
+  bool index_free_;
+
+  std::unique_ptr<LocalDocumentsView> local_documents_;
+  std::unique_ptr<WrappedMutationQueue> mutation_queue_;
+  std::unique_ptr<WrappedRemoteDocumentCache> remote_documents_;
+
+  int mutations_read_by_query_ = 0;
+  int mutations_read_by_key_ = 0;
+  int documents_read_by_query_ = 0;
+  int documents_read_by_key_ = 0;
+};
+
+/** A MutationQueue that counts document reads. */
+class WrappedMutationQueue : public MutationQueue {
```

Comment:
Port of https://github.com/firebase/firebase-android-sdk/blob/2a2e789629a7dad13e1007b6cdec26c96e73defe/firebase-firestore/src/test/java/com/google/firebase/firestore/local/CountingQueryEngine.java#L154